IBX-10701: `ibexa_tracking_script` function described in Dev-doc (#3105)
ezsystems/developer-documentation: docs/cdp/raptor_integration/tracking_script.md

---
description: Integrate the tracking script to collect user interactions.
edition: experience
---

# Tracking script

`ibexa_tracking_script` function allows to embed main tracking script into the website.
It loads the initial script into `window.raptor`, enabling event tracking, for example, page visits, product views, or buys, from the frontend.
It can be overridden in multiple ways to support custom implementations and to render code snippet through [[= product_name_base =]] Design Engine.

Tracking can be conditionally initialized depending on cookie consent logic.
By default, the function returns the script for client-side use, while it can return nothing when used server-side.

## Embedd tracking script

To enable tracking, tracking script must be embedded into the website’s layout.

> To embedd tracking script, add the twig function `ibexa_tracking_script()` into the <head> section of your base layout template: `layout.html.twig`.

Reviewer: I'd emphasize here that `layout.html.twig` is an example. Hard to say how Partners name their templates.

This function accepts following parameters:

|Parameter     |Type   |Default value                |Remarks                                   |
|--------------|-------|-----------------------------|------------------------------------------|
|`customerId`  |string |From SiteAccess configuration|Can be overridden for custom customer IDs.|

> |`hasConsented`|boolean|false                        |Controls loading of tracking based on user consent at render time.|

Reviewer: I'd add emphasized information: pass true only if you're certain during backend rendering that user has consented to tracking.

Author: Described below:

_If `hasConsented` is set to `true` in the template, the tracking script is initialized automatically.
This value should be set if user consent for tracking cookies is already known at render time.
If `hasConsented` parameter is set to `false`, tracking should be enabled by dispatching the `enableTracking` CustomEvent after consent is granted, for example through a custom cookie script in layout._

Example setup:

``` bash
{{ ibexa_tracking_script(customerId: '123', hasConsented: true) }}

> ```

Reviewer: I'd recommend by default 
``` bash
{{ ibexa_tracking_script() }}
```
and the current example with all the options as a second example.

> ---

Reviewer 2: Note: this file is not added to mkdocs.yaml

Reviewer 3: @mnocon @julitafalcondusza I think #3105 and #3110 should be merged into one page that could be used at least in

- mkdocs.yaml
- [Twig function reference cards](https://doc.ibexa.co/en/5.0/templating/twig_function_reference/twig_function_reference/#twig-function-reference)
- [raptor_connector.md cards (already kind of done)](https://github.com/ibexa/documentation-developer/pull/3101/changes#diff-4393819c129ff102895a1bdd9e78884089063745bc2efec9d5842d7cd47f0be4R25-R30)
- [connector_configuration.md when quoting the functions in the `client` setting](https://github.com/ibexa/documentation-developer/pull/3101/changes#diff-ef8a7f940b1b51e29839b2a69de2f8a93b3abcc5721b3307e90ae72e531eb860R30)

Author: All PRs will be merged in one branch, then I'll think about connecting pages :)

> description: Integrate the tracking script to collect user interactions.

Reviewer 3: ```suggestion
description: Integrate the tracking script to collect user interactions.
month_change: true
```
Edit: LTS Update removed

---

# Tracking script

> `ibexa_tracking_script` function allows to embed main tracking script into the website.

Reviewer 2: ```suggestion
`ibexa_tracking_script` Twig function allows you to embed main tracking script into the website.
```

> `ibexa_tracking_script` function allows to embed main tracking script into the website.

Reviewer 2: I'd add a section to the Twig Reference as well - Raptor Twig functions, with this entry linking here for more details.

Author: Will be added

> It loads the initial script into `window.raptor`, enabling event tracking, for example, page visits, product views, or buys, from the frontend.

Reviewer 2: ```suggestion
It loads the initial script into `window.raptor`, enabling tracking events, for example, page visits, product views, or buys, from the frontend.
```

> It can be overridden in multiple ways to support custom implementations and to render code snippet through [[= product_name_base =]] Design Engine.

Reviewer 2: Let's add a link to https://doc.ibexa.co/en/5.0/templating/design_engine/design_engine/

Tracking can be conditionally initialized depending on cookie consent logic.
By default, the function returns the script for client-side use, while it can return nothing when used server-side.

## Embedd tracking script

To enable tracking, tracking script must be embedded into the website’s layout.
To embedd tracking script, add the twig function `ibexa_tracking_script()` into the <head> section of your base layout template, for example, `layout.html.twig`.

> This function accepts following parameters:

Reviewer 2: ```suggestion
This function accepts the following parameters:
```

|Parameter     |Type   |Default value                |Remarks                                   |
|--------------|-------|-----------------------------|------------------------------------------|
|`customerId`  |string |From SiteAccess configuration|Can be overridden for custom customer IDs.|
|`hasConsented`|boolean|false                        |Controls loading of tracking based on user consent at render time.|

Default setup:

``` bash
{{ ibexa_tracking_script() }}
```
Example setup using parameters:

``` bash
{{ ibexa_tracking_script(customerId: '123', hasConsented: true) }}
```

Without setting custom `customerId` parameter, the function renders the tracking script using the configured `customerID` from the [connector configuration](raptor_connector.md#configuration).
It can be overridden by providing a custom value if needed.

If `hasConsented` is set to `true` in the template, the tracking script is initialized automatically.
This value should be set if user consent for tracking cookies is already known at render time.

> If `hasConsented` parameter is set to `false`, tracking should be enabled by dispatching the `enableTracking` CustomEvent after consent is granted, for example through a custom cookie script in layout.

Reviewer 2: ```suggestion
If `hasConsented` parameter is set to `false`, tracking should be enabled by dispatching a custom JavaScript event after consent is granted, for example through a custom script in layout.
```

This sentence and the one below are very similar together - usually the first sentence should be an intrudction and the second one should have more details, here it's mixed - the first sentence already has all the info, the second only repeats (with less details than the first one).

Please have a look, maybe you see another way of rewriting this

> The recommended method to integrate the tracking script with custom front-end logic is to dispatch the `enableTracking` JavaScript event after tracking cookie consent is granted:

Reviewer 2: Inconsistent frontend spelling (front-end also occurs on this page)

Author: I've just checked our doc - let's keep "front-end", fixed in line 8 then.

> ``` bash

Reviewer 2: ```suggestion
``` js
```

document.dispatchEvent(new CustomEvent('enableTracking'));
```

!!! note

>     In debug (development) mode, the provided script outputs diagnostic information to the console.

Reviewer 2: ```suggestion
    In [Symfony's debug mode](https://symfony.com/doc/current/reference/configuration/kernel.html#kernel-debug), the provided script outputs diagnostic information to the console.
```
(please use Symfony doc variable)

    This output is not included in production environment.

## Complex integration

> For more complex integrations, the [[= product_name_base =]] Design Engine can be used to override parts or entire templates that render the tracking script.

Reviewer 2: I'd add a link to https://doc.ibexa.co/en/5.0/templating/design_engine/design_engine/

Author: Added in line 9 (it's enough once per page)

> |Template                                       |Example project path                                       |

Reviewer 2: Would it make sense to mention the variables passed to the templates, to know what's available?

Reviewer: > Would it make sense to mention the variables passed to the templates, to know what's available?

Good idea. See the next diff comments.

|-----------------------------------------------|-----------------------------------------------------------|

> |`@ibexadesign/ibexa/tracking/script.html.twig` |`templates/themes/standard/ibexa/tracking/script.html.twig`|

Reviewer: Variables:

- `customer_id` (string),
- `script_url` (string) by default `//deliver.raptorstatic.com/script/raptor-3.0.min.js`, can be configured via `ibexa.connector.raptor.tracking_script.url` Symfony Dependency Injection container parameter (not SiteAccess-aware).
- `has_consented` (boolean) - by default false, unless explicitly passed as function argument
- `debug` (boolean) - `kernel.debug` Symfony dependency injection container parameter, usually true in `dev` enviroment and false in `prod` enviroment, respectively.

other globals are also included, but not sure if it makes sense to mention them and/or describe, e.g.,`app` or `ibexa`.

The default template also provides Twig block `ibexa_tracking_script`, which includes `script.js.twig`. If the template is extended, the block can be overridden to customize behavior.

> |`@ibexadesign/ibexa/tracking/script.js.twig`   |`templates/themes/standard/ibexa/tracking/script.js.twig`  |

Reviewer: Variables:

- `customer_id` (string)
- `script_url` (string)
- `debug`
passed down from the above script

> It's possible to extend `script.html.twig` by combining the [[= product_name_base =]] Design Engine with standard Symfony template reference in `./templates/themes/standard/ibexa/tracking/script.html.twig`:

Reviewer 2: ```suggestion
It's possible to extend `script.html.twig` by combining the [[= product_name_base =]] Design Engine with standard Symfony template reference in `templates/themes/standard/ibexa/tracking/script.html.twig`:
```
For consistenct with the other paths.

``` html+twig
{% extends '@IbexaConnectorRaptor/themes/standard/ibexa/tracking/script.html.twig' %}
{% block ibexa_tracking_script %}
    console.log('My custom tracking script, but relying on loadTracking function.');
{% endblock %}
```

> In most cases, the preferred approach is to do the opposite:

Reviewer 2: I think it would be nice to explain what the opposite approach is - instead of extending the script (replacing the internal part of the script), here we're wrapping the script with our custom logic from the outside?

Reviewer: > I think it would be nice to explain what the opposite approach is - instead of extending the script (replacing the internal part of the script), here we're wrapping the script with our custom logic from the outside?

You can either override outer template and include inner Raptor script that comes from Raptor documentation, or you can use outer template but override inner if you prefer your Raptor script to look differently - you're using your own approach, Or you can override both.
This is vague because custom implementations around consent are not known to me nor handled by Ibexa DXP.

Not sure if this answers your question :)

> ``` bash

Reviewer 2: ```suggestion
``` html+twig
```

<script type="text/javascript">
    if (myCustomConsentIsGiven) {
           {{ include('@ibexadesign/ibexa/tracking/script.js.twig', {'customer_id': customer_id}) }}
    }
</script>
```

### Example custom integration

Example custom integration with [TermsFeed](https://www.termsfeed.com/):

> ``` bash

Reviewer 2: ```suggestion
``` html
```

---
description: Integrate the tracking script to collect user interactions.
edition: experience
---

# Tracking script

`ibexa_tracking_script` function allows to embed main tracking script into the website.
It loads the initial script into `window.raptor`, enabling event tracking, for example, page visits, product views, or buys, from the frontend.
It can be overridden in multiple ways to support custom implementations and to render code snippet through [[= product_name_base =]] Design Engine.

Tracking can be conditionally initialized depending on cookie consent logic.

> By default, the function returns the script for client-side use, while it can return nothing when used server-side.

Reviewer 2: I don't understand this sentence - I thought this function is for website (frontend) tracking, not sure what is meant here when mentioning server-side.

Reviewer: > I don't understand this sentence - I thought this function is for website (frontend) tracking, not sure what is meant here when mentioning server-side.

@mnocon that was initially the case, which during QA proved to have very bad DX. Essentially, it would require Partner to implement custom event subscribers to perform server-side tracking. The current idea is that you transparently use both Twig functions in your templates: `ibexa_tracking_script` and `ibexa_tracking_track_event` which behaves differently depending if you configured server-side or client.

To answer the first question that might pop up - why can't we implement subscribers ourselves? In some cases - like during buy process - yes, but in general loading items does not yield tracking action - only the request to display item(s) does. Which are done via custom templates, hence default implementation is hard to provide otherwise.

Reviewer 2: Ok, I see that this topic is explained a bit in https://github.com/ibexa/documentation-developer/pull/3110/changes, which I haven't reviewed yet - that's why I'm missing some context.

I'm discussing right now with Julita if these 2 pages should be mentioned together, because the explanation just in this PR is not clear for me

> ## Embedd tracking script

Reviewer 2: Vale is right

To enable tracking, tracking script must be embedded into the website’s layout.

> To embedd tracking script, add the twig function `ibexa_tracking_script()` into the <head> section of your base layout template, for example, `layout.html.twig`.

Reviewer 2: Vale is right

> To embedd tracking script, add the twig function `ibexa_tracking_script()` into the <head> section of your base layout template, for example, `layout.html.twig`.

Reviewer 2: ```suggestion
To embedd tracking script, add the twig function `ibexa_tracking_script()` into the <head> section of your base layout template, for example, `pagelayout.html.twig`.
```

Let's use the name of the file we provide by default:
https://github.com/ibexa/recipes-dev/blob/master/ibexa/oss/5.0/templates/themes/standard/pagelayout.html.twig

for consistency

This function accepts following parameters:

|Parameter     |Type   |Default value                |Remarks                                   |
|--------------|-------|-----------------------------|------------------------------------------|

> |`customerId`  |string |From SiteAccess configuration|Can be overridden for custom customer IDs.|

Reviewer 2: So this is something like "Raptor account ID", it has nothing to do with the customer of the website that we're tracking?

I'd add that info to the table to make it clearer - when reading this, I was thinking that customer = user, visitor of the website.

```suggestion
|`customerId`  |string |From SiteAccess configuration|Raptor account ID. Can be overridden for custom customer IDs.|
```

Reviewer 2: Reading https://github.com/ibexa/documentation-developer/pull/3101/changes, I'd just link to this section from here:

```
- `customer_id` - an identifier used to authenticate requests to the Recommendation Engine. This value is can be found as `Account number` in [Raptor Control Panel](https://controlpanel.raptorsmartadvisor.com/).
```
```suggestion
|`customerId`  |string |From SiteAccess configuration|Raptor account ID. Can be overridden for custom customer IDs.|
```

|`hasConsented`|boolean|false                        |Controls loading of tracking based on user consent at render time.|

Default setup:

> ``` bash

Reviewer 2: ```suggestion
``` html+twig
```
For better rendering

{{ ibexa_tracking_script() }}
```
Example setup using parameters:

> ``` bash

Reviewer 2: ```suggestion
``` html+twig
```
